fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java

```diff
@@ -3087,8 +3091,123 @@ private void setOperationContextProperties(FHIROperationContext operationContext
+        if (indexIds == nulll || indexIds.isEmpty()) {
+            throw new IllegalArgumentExecption("No indexIds sent to the $reindex list method");
+        }
+
+        /*
+         * How the backoff works...
+         * indexIds[1,2,3,4,5,6,7,8,9,10]
+         *
+         * Pass 1 left=0, right=10, max=10
+         * Result: Deadlock
+         *
+         * Pass2 left=0, right=1, max=10
+         * Move over by 1
+         * Result: Pass
+         *
+         * Pass3 left=1, right=2, max=10
+         * Move over by 1
+         * Result: Pass
+         *
+         * ... all the way up to 10
+         *
+         * Return total count back to caller.
+         *
+         * @implNote tried divide and conquer and it caused it to try large pass fail, small pass succeed, and therefore chose a small linear pass.
+         */
+
+        // Maximum number to retry
+        final int TX_ATTEMPTS = 5;
+
+        int result = 0;
+        int max = indexIds.size();
+
+        int left = 0;
+        int right = max;
+
+        int window = 0;
+        while (left < max) {
+            window++;
+            if (log.isLoggable(Level.FINE)) {
+                log.fine("$reindex window[" + window + "] -> left=[" + left + "] right=[" + right + "] max=[" + max + "]");
+            }
+
+            boolean backoff = false;
+            List<Long> subListIndexIds = indexIds.subList(left, right);
+
+            int attempt = 1;
+            while (attempt <= TX_ATTEMPTS && !backoff) {
+                FHIRTransactionHelper txn = new FHIRTransactionHelper(getTransaction());
+                txn.begin();
+                try {
+                    FHIRPersistenceContext persistenceContext = null;
+                    result += persistence.reindex(persistenceContext, operationOutcomeResult, tstamp, subListIndexIds, null);
+                    attempt = TX_ATTEMPTS + 1; // end the retry loop
+                } catch (FHIRPersistenceDataAccessException x) {
+                    // At this point, the transaction is marked for rollback
+                    if (x.isTransactionRetryable()) {
+                        if (x.getCause() instanceof LockException && ((LockException) x.getCause()).isDeadlock()) {
+                            backoff = true;
+                            long wait = RANDOM.nextInt(5000);
+                            log.info("attempt #" + window + "/" + attempt + " failed, retrying transaction, backing off [wait=" + wait + "ms]");
+                            Thread.sleep(wait);
+                        } else {
+                            log.info("attempt #" + window + "/" + attempt + " failed, retrying transaction, not backing off");
+                        }
+                    } else {
+                        throw x;
+                    }
+                } finally {
+                    txn.end();
+                }
+                attempt++;
+            }
+
+            if (backoff) {
```

Comment:
If I'm understanding this right, if the failed attempts for a subset are isTransactionRetryable(), this will continue to try to process more indexIds, instead of giving up, and return the number that were successful, even though some of them failed.  Is that ok?

Reply:
they have all failed, and so we switch to processing them one at a time.